src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py

```diff
@@ -79,66 +71,93 @@ class KalmanODESolution(ODESolution):
-        dy_rvs = [
-            _project_rv(self.proj_to_dy, rv) for rv in self.kalman_posterior.state_rvs
-        ]
-        return pnrv_list._RandomVariableList(dy_rvs)
-
-    def __call__(
-        self, t: float
-    ) -> typing.Union[pnrv.RandomVariable, pnrv_list._RandomVariableList]:
-        out_rv = self.kalman_posterior(t)
-
-        if np.isscalar(t):
-            return _project_rv(self.proj_to_y, out_rv)
-
-        return pnrv_list._RandomVariableList(
-            [_project_rv(self.proj_to_y, rv) for rv in out_rv]
+        states = _randomvariablelist._RandomVariableList(
+            [_project_rv(self.proj_to_y, rv) for rv in self.kalman_posterior.states]
+        )
+        derivatives = _randomvariablelist._RandomVariableList(
+            [_project_rv(self.proj_to_dy, rv) for rv in self.kalman_posterior.states]
+        )
+        super().__init__(
+            locations=kalman_posterior.locations, states=states, derivatives=derivatives
         )
 
+    def interpolate(self, t: FloatArgType) -> random_variables.RandomVariable:
+        out_rv = self.kalman_posterior.interpolate(t)
+
+        return _project_rv(self.proj_to_y, out_rv)
+
     def sample(
         self,
-        t: typing.Optional[float] = None,
-        size: typing.Optional[probnum.type.ShapeArgType] = (),
+        t: Optional[DenseOutputLocationArgType] = None,
+        size: Optional[ShapeArgType] = (),
+        random_state: Optional[RandomStateArgType] = None,
     ) -> np.ndarray:
-        """Sample from the Gaussian filtering ODE solution by sampling from the Gauss-
-        Markov posterior."""
-        size = probnum.utils.as_shape(size)
+        """Sample from the Gaussian filtering ODE solution."""
+        size = utils.as_shape(size)
+        t = np.asarray(t) if t is not None else None
+
+        # If self.locations are used, the final RV in the list is informed
+        # about all the data. If not, the final data point needs to be
+        # included in the joint sampling, hence the (len(t) + 1) below.
+        if t is None:
+            t_shape = (len(self.locations),)
+        else:
+            t_shape = (len(t) + 1,)
+
+        # Note how here, the dimension of the underlying Kalman posterior
+        # is used in order to generate the base measure samples.
+        # This is because KalmanODESolution essentially delegates
+        # everything to KalmanPosterior
+        rv_list_shape = (len(self.kalman_posterior.states[0].mean),)
+
+        base_measure_realizations = stats.norm.rvs(
+            size=(size + t_shape + rv_list_shape), random_state=random_state
+        )
+        return self.transform_base_measure_realizations(
+            base_measure_realizations=base_measure_realizations, t=t, size=size
+        )
+
+    def transform_base_measure_realizations(
+        self,
+        base_measure_realizations: np.ndarray,
+        t: Optional[DenseOutputLocationArgType] = None,
+        size: Optional[ShapeArgType] = (),
```

Comment:
it should be possible to remove this argument altogether. Size should be possible to infer from the shape of base_measure_realizations?!